src/state/transactionState.ts

```diff
@@ -0,0 +1,560 @@
+  const pendingTransactionIdx = txGroup.findIndex(
+    (tx) => tx.status === TransactionStatus.Pending,
+  );
+  if (pendingTransactionIdx > -1) return pendingTransactionIdx;
+  return txGroup.findIndex(
+    (tx) =>
+      tx.status === TransactionStatus.Ready ||
+      tx.status === TransactionStatus.Failed,
+  );
+};
+
+// Get transaction values to show in title or description
+export const getGroupValues = <T>(
+  txGroup: T[], // For now, just returns the first transaction if we have one
+) => txGroup[0];
+
+export const getGroupId = (txGroup: TransactionType[]) => txGroup[0].group.id;
+
+export const getGroupKey = (txGroup: TransactionType[]) => txGroup[0].group.key;
+
+export const findTransactionGroupByKey = (
+  txGroups: TransactionType[][],
+  key: string,
+) => txGroups.find((txGroup) => txGroup[0].group.key === key);
+
+// Get count of all transactions in the redux store
+export const transactionCount = (transactions: TransactionType[]) =>
+  transactions.flatMap((tx) => tx).length;
+
+export const useGroupedTransactions = () => {
+  const { user } = useAppContext();
+  const userAddress = utils.getAddress(user?.walletAddress as string);
+  const { data, loading } = useGetUserTransactionsQuery({
+    variables: {
+      userAddress,
+      limit: TX_PAGE_SIZE,
+    },
+    skip: !user,
+  });
+
+  const transactions = useMemo(() => {
+    const rawTransactions =
+      data?.getTransactionsByUser?.items.map(convertTransactionType) || [];
+    return filter(
+      mapValues(
+        // Group transactions by groupId
+        groupBy(rawTransactions, 'groupId'),
+        // Map each group to be sorted by group.index
+        (group) => orderBy(group, [(item) => item?.group?.index], ['asc']),
+      ),
+      // Filter all items that do not contain a transaction of index 0
+      // (this means they were cut off within the group)
+      (group) => !!group.find((item) => item?.group?.index === 0),
+      // Finally we sort the resulting array of arrays by the createdAt value of the first transaction within a group (descending)
+    ).sort(
+      (groupA, groupB) =>
+        new Date(groupB[0]?.createdAt || 0).valueOf() -
+        new Date(groupA[0]?.createdAt || 0).valueOf(),
+    );
+  }, [data?.getTransactionsByUser?.items]);
+
+  const groupState = getGroupStatusAll(transactions, loading);
+
+  return { groupState, transactions, loading };
+};
+
+export const getTransaction = async (id: string, fetchPolicy?: FetchPolicy) => {
+  // Get the transaction from apollo
+  const apolloClient = getContext(ContextModule.ApolloClient);
+
+  const { data } = await apolloClient.query<
+    GetTransactionQuery,
+    GetTransactionQueryVariables
+  >({
+    query: GetTransactionDocument,
+    variables: { id },
+    fetchPolicy,
+  });
+
+  if (!data.getTransaction) {
+    throw new Error('Transaction not found.');
+  }
+  return convertTransactionType(data.getTransaction);
+};
+
+export const fetchTransaction = async ({ id }) => {
```

Comment:
do we need both this and `getTransaction`?

Reply:
That is a very good point, let me look into this.